overlord/hookstate/hookmgr.go

```diff
@@ -84,6 +84,31 @@ type HookSetup struct {
 // Manager returns a new HookManager.
 func Manager(s *state.State) (*HookManager, error) {
 	runner := state.NewTaskRunner(s)
+
+	// Make sure we only run 1 hook task for given snap at a time
+	runner.SetBlocked(func(thisTask *state.Task, running []*state.Task) bool {
+		// check if we're a hook task, probably not needed but let's take extra care
+		if thisTask.Kind() != "run-hook" {
+			return false
+		}
+		var hooksup HookSetup
+		if thisTask.Get("hook-setup", &hooksup) != nil {
+			return false
+		}
+		thisSnapName := hooksup.Snap
+		// examine all hook tasks, block thisTask if we find any other hook task affecting same snap
+		for _, t := range running {
+			if t.Kind() != "run-hook" || t.Get("hook-setup", &hooksup) != nil {
```

Comment:
because Snap is something that is always set in HookSetup this not an actual problem here but in general because json.Unmarshal updates without resetting target values,  doing Get on a reused variable can be risky